From afcb3056116b0a45521820ebd8a214c46418a5a1 Mon Sep 17 00:00:00 2001
From: "Markus L. (FivePB)" <admin@fivepb.me>
Date: Mon, 9 Nov 2020 09:47:50 +0100
Subject: [PATCH] Preliminary 1.17 support


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
index d2a11a82..2a60769d 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
@@ -245,13 +245,16 @@ public enum Protocol
                     map( ProtocolConstants.MINECRAFT_1_17, 0x59 )
             );
             TO_CLIENT.registerPacket( ClearTitles.class,
+                    ClearTitles::new, // Waterfall - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_17, 0x10 )
             );
             TO_CLIENT.registerPacket(
                     Subtitle.class,
+                    Subtitle::new, // Waterfall - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_17, 0x57 )
             );
             TO_CLIENT.registerPacket( TitleTimes.class,
+                    TitleTimes::new, // Waterfall - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_17, 0x5A )
             );
             TO_CLIENT.registerPacket(
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index 8c3a78d6..6500f6bb 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -34,7 +34,8 @@ public class ProtocolConstants
     public static final int MINECRAFT_1_16_2 = 751;
     public static final int MINECRAFT_1_16_3 = 753;
     public static final int MINECRAFT_1_16_4 = 754;
-    public static final int MINECRAFT_1_17 = 1073741852;
+    public static final int MINECRAFT_1_17 = 755; // Waterfall
+    public static final int SNAPSHOT_PROTOCOL = 1073741852; // Waterfall
     public static final List<String> SUPPORTED_VERSIONS;
     public static final List<Integer> SUPPORTED_VERSION_IDS;
 
@@ -84,7 +85,12 @@ public class ProtocolConstants
         if ( SNAPSHOT_SUPPORT )
         {
             supportedVersions.add( "1.17.x" );
-            supportedVersionIds.add( ProtocolConstants.MINECRAFT_1_17 );
+            // Waterfall start
+            supportedVersionIds.add(
+                    ProtocolConstants.MINECRAFT_1_17,
+                    ProtocolConstants.SNAPSHOT_PROTOCOL
+            );
+            // Waterfall end
         }
 
         SUPPORTED_VERSIONS = supportedVersions.build();
-- 
2.30.1 (Apple Git-130)

